modules/images/dominant-color/load.php

```diff
@@ -0,0 +1,375 @@
+
+
+/**
+ * Add the dominant color metadata to the attachment.
+ *
+ * @param array $metadata Metadata for the attachment.
+ * @param int   $attachment_id attachement id.
+ *
+ * @return array $metadata
+ * @since n.e.x.t
+ */
+function dominant_color_has_transparency_metadata( $metadata, $attachment_id ) {
+
+	if ( ! wp_attachment_is_image( $attachment_id ) ) {
+		return $metadata;
+	}
+
+	$has_transparency = dominant_color_get_has_transparency( $attachment_id );
+
+	if ( ! empty( $has_transparency ) ) {
+		$metadata['has_transparency'] = $has_transparency;
+	} else {
+		$metadata['has_transparency'] = false;
+	}
+
+	return $metadata;
+}
+
+add_filter( 'wp_generate_attachment_metadata', 'dominant_color_has_transparency_metadata', 10, 2 );
+
+/**
+ * Filter various image attributes to add the dominant color to the image
+ *
+ * @param array  $attr Attributes for the image markup.
+ * @param object $attachment Image attachment post.
+ *
+ * @return mixed
+ * @since n.e.x.t
+ */
+function dominant_color_tag_add_adjust_to_image_attributes( $attr, $attachment ) {
+
+	$image_meta = wp_get_attachment_metadata( $attachment->ID );
+	if ( ! is_array( $image_meta ) ) {
+		return $attr;
+	}
+
+	$has_transparency = isset( $image_meta['has_transparency'] ) ? $image_meta['has_transparency'] : false;
+
+	$extra_class = '';
+	if ( ! isset( $attr['style'] ) ) {
+		$attr['style'] = ' --has-transparency: ' . $has_transparency . '; ';
+	} else {
+		$attr['style'] .= ' --has-transparency: ' . $has_transparency . '; ';
+		$extra_class    = ' has-transparency ';
+	}
+
+	if ( isset( $image_meta['dominant_color'] ) ) {
+
+		$attr['data-dominant-color'] = $image_meta['dominant_color'];
+
+		if ( empty( $attr['style'] ) ) {
+			$attr['style'] = '';
+		}
+		$attr['style'] .= '--dominant-color: #' . $image_meta['dominant_color'] . ';';
+
+		$extra_class .= ( dominant_color_color_is_light( $image_meta['dominant_color'] ) ) ? 'dominant-color-light' : 'dominant-color-dark';
+		if ( empty( $attr['class'] ) ) {
+			$attr['class'] = '';
+		}
+		if ( isset( $attr['class'] ) && ! array_intersect( explode( ' ', $attr['class'] ), explode( ' ', $extra_class ) ) ) {
+			$attr['class'] .= ' ' . $attr['class'];
+		}
+	}
+
+	return $attr;
+}
+
+add_filter( 'wp_get_attachment_image_attributes', 'dominant_color_tag_add_adjust_to_image_attributes', 10, 2 );
+
+/**
+ * Filter image tags in content to add the dominant color to the image.
+ *
+ * @param string $filtered_image The filtered image.
+ * @param string $context The context of the image.
+ * @param int    $attachment_id The attachment ID.
+ *
+ * @return string image tag
+ * @since n.e.x.t
+ */
+function dominant_color_tag_add_adjust( $filtered_image, $context, $attachment_id ) {
+
+	$image_meta = wp_get_attachment_metadata( $attachment_id );
+
+	if ( ! is_array( $image_meta ) || ! isset( $image_meta['dominant_color'], $image_meta['has_transparency'] ) ) {
+
+		return $filtered_image;
+	}
```

Comment:
Here, we look up the image metadata before calling the filter below. This is not a problem in itself, but the existing core filters for similar `img` modifications work the other way around:
1. Call the filter.
2. If the filter says we should adjust the `img`, run the necessary logic.

In other words, I think we should reverse this and first call the filter below. If it doesn't return `true`, this function shouldn't do anything. Only if it returns `true`, we should get the attachment metadata. And only if then those values are set as needed, we should make the adjustments.